### PR TITLE
create resources as root to make multiple deployments possible

### LINKED ISSUE #10: Add a syntax that allows grouping of multiple resources
Currently it is only possible to define one resource per script. This however is not very comfortable. So there needs to be a language construct that lets one do exactly that. The `ResourcesBuilder` aims to solve this issue.

```kotlin
resources {
    // Resource definitions go here ...
}
```

### PR DESCRIPTION
**Pull request by Tim-Obert**
add publish block to publish ports and domains
add loadBalancer block
Closes #10

### CODE REVIEW

File: src/script/data/Domain.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import kotlinx.serialization.Serializable
+import org.valiktor.functions.isNotBlank
+import org.valiktor.functions.isNotEmpty
+import org.valiktor.functions.matches
+import org.valiktor.validate
+
+@Serializable
+data class Domain(
+    val host: String
```
Comment (Tim-Obert) [inline src/script/data/Domain.kt:11]: maybe we need a better naming there

File: src/script/data/PublishDomains.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package it.oechsler.script.data
+
+import kotlinx.serialization.Serializable
+import org.valiktor.functions.isIn
+import org.valiktor.validate
+
+@Serializable
+data class PublishDomains(
+    var port: Int,
+    var domains: Set<Domain>
```
Comment (Tim-Obert) [inline src/script/data/PublishDomains.kt:10]: Not sure with the naming here. The current syntax would create sth. like:
` "publish": {
                "ports": [
                    {
                        "port": 8080
                    }
                ],
                "domains": [
                            {
                                "host": "oechsler.it"
                            }
                        ]
                    }
                ]
            }`

File: src/script/data/Resources.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+data class Resources(
+    val deployments: Set<Deployment>,
+    val loadBalancers: Set<LoadBalancer>,
+) {
+    init {
+        validate(this) {
+
+        }
+    }
+}
```
Comment (oechsler) [inline src/script/data/Resources.kt:16]: This can be removed ...

File: src/script/language/DeploymentBuilder.kt [hunk 4]
```diff
@@ -20,13 +21,14 @@ class DeploymentBuilder private constructor(val name: String) : ScriptRoot {
...
+    private lateinit var publish: Publish
 
     fun tags(vararg tags: String) {
         this.tags = tags.toSet()
     }
 
     fun toDeployment(): Deployment {
-        return Deployment(name, tags, replicas, containers)
+        return Deployment(name, tags, replicas, containers, publish)
     }
```
Comment (oechsler) [inline src/script/language/DeploymentBuilder.kt:32]: Maybe put this to the end off the class for consistency reasons

File: src/script/language/PublishDomainBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    fun domain(domain: String) {
+        val mutableSet = this.domainsToPublish.toMutableSet()
+        mutableSet.add(Domain(domain))
+        this.domainsToPublish = mutableSet.toSet()
+    }
+
+
+    fun toPublishDomains(): PublishDomains {
+
```
Comment (oechsler) [inline src/script/language/PublishDomainBuilder.kt:26]: ?

File: src/script/language/ResourcesBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    fun loadBalancer(name: String, block: LoadBalancerBuilder.() -> Unit) {
+        val mutableSet = this.loadBalancers.toMutableSet()
+        mutableSet.add(LoadBalancerBuilder.loadBalancer(name, block).toLoadBalancer())
+        this.loadBalancers = mutableSet
+    }
+
+    override fun apply() {
+        val serializer = Json { prettyPrint = true }
+        println(serializer.encodeToString(toResources()))
+    }
```
Comment (oechsler) [inline src/script/language/ResourcesBuilder.kt:42]: This should only process the other `ScriptRoot` instances that it holds

**Review by oechsler — CHANGES_REQUESTED** [review]

File: src/script/data/Resources.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package it.oechsler.script.data
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Resources(
+    val deployments: Set<Deployment>,
+    val loadBalancers: Set<LoadBalancer>,
+)
```
Comment (oechsler) [inline src/script/data/Resources.kt:9]: Isn't it possible to remove this now?

**Review by oechsler — COMMENTED** [review]
Make sure that if `Resources.kt` is not needed anymore to remove it. Other than that, this PR is approved!

**Review by oechsler — APPROVED** [review]

**Pull request merged**